Fix personal setting for apps which are not whitelisted

sharidas opened the pull request:
When apps are not whitelisted for guest user
they should not be visible in the personal settings.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
When apps are not enabled for guest users, they should not be listed in personal settings. In this change a new method is introduced in the user class `getExtendedAttributes()`. The purpose of this method is to grab the attributes related to the user for the apps. So when this method is invoked an event is emitted which is listened by the apps, and populate the event array with the attributes in the form of key => value. So when logged in as guest user the guest app would populate the event array with `isGuest` and `whiteListedApps`. Which could be used by core or any other apps. This feature is used SettingsManagers `loadPanel` to filter out the apps which are not whitelisted.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Do not show the apps which are not whitelisted for the guest user.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Enable customgroups. Make sure its not listed as whitelisted apps in guest configuration.
- As guest user navigate to the settings page.
- The settings page should not show customgroups. And it does not.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](../changelog/README.md)

sharidas committed:
Get extended attributes for user for apps

Get extended attributes for user from the apps. A
new method which emits events to get the attributes
for user, which is listened by apps to provide the
attributes for the specific user. Use case for example:
When apps are not whitelisted for guest user
they should not be visible in the personal settings.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

jvillafanez commented on lib/private/User/User.php:
```diff
@@ -555,4 +555,21 @@ public function setSearchTerms(array $terms) {
+			$this->eventDispatcher = \OC::$server->getEventDispatcher();
+		}
+
+		$event = new GenericEvent('appAttributesForUser', []);
```
Use a custom event so we can have full control on how we want to store the information

sharidas commented on lib/private/App/AppManager.php:
```diff
@@ -160,9 +160,20 @@ public function getInstalledApps() {
-		$appsForUser = \array_filter($apps, function ($enabled) use ($user) {
-			return $this->checkAppForUser($enabled, $user);
-		});
+		$whiteListedAppsForGuest = [];
```
Moved to the `checkAppForUser` method.

sharidas commented on lib/private/App/AppManager.php:
```diff
@@ -182,7 +193,15 @@ public function isEnabledForUser($appId, $user = null) {
 		$installedApps = $this->getInstalledAppsValues();
 		if (isset($installedApps[$appId])) {
-			return $this->checkAppForUser($installedApps[$appId], $user);
+			$whiteListedApps = [];
```
Moved to `checkAppForUser` method.

sharidas commented on lib/private/App/AppManager.php:
```diff
@@ -191,9 +210,14 @@ public function isEnabledForUser($appId, $user = null) {
 	 * @return bool
 	 */
-	private function checkAppForUser($enabled, $user) {
+	private function checkAppForUser($enabled, $user, $appName = null, $whiteListedAppsForGuest = []) {
```
Removed the whitelist and kept the appName.

jvillafanez commented on lib/private/User/User.php:
```diff
@@ -555,4 +556,23 @@ public function setSearchTerms(array $terms) {
+	 */
+	public function getExtendedAttributes() {
+		if ($this->eventDispatcher === null) {
+			$this->eventDispatcher = \OC::$server->getEventDispatcher();
```
I don't like this. This should be done in the constructor, not here. Nobody will expect that the server's event dispatcher will be assigned in a "get" method.
If you don't want to throw an exception, returning "false" could be fine. (whatever you decide, make sure is documented in the interface)

sharidas replied:
Correct. We can safely remove this. I was bit worried when I saw this line in the file https://github.com/owncloud/core/blob/master/lib/private/User/User.php#L425. I have tested it by removing it in my instance and it works. So in the next version, it won't be there.

jvillafanez replied:
The event dispatcher can be null (for a reason I don't understand). This is something we need to handle, unless we force the event dispatcher to be not null in the constructor.

sharidas commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,73 @@
+	 * @param string $appName
+	 * @param array $data
+	 */
+	public function setAttributes($app, $data) {
```
> How are we going to deal with multiple apps? I think we could have a "first in wins" policy: if the attribute doesn't exists, write it and return false; it it already exists, ignore it and return false

There is another possibile flaw I see here, which is say for example if we use the policy to only add new attributes which don't exist and leave if the attribute already exist. Then it could also happen for example the user updated the app attribute and we might miss that, because we don't update. We only add the new attributes and skip if the attribute already exists. imho, shouldn't we consider to add/update it anyway? Or did I misread your feedback, please let me know your opinion.

sharidas commented on lib/private/App/AppManager.php:
```diff
@@ -191,9 +191,23 @@ public function isEnabledForUser($appId, $user = null) {
+			/**
+			 * Check if the user is guest user, if so get the apps which it is allowed for.
+			 */
+			if (\array_key_exists('guests', $userAppAttributes)) {
```
- If `guests` and `whitelistedApps` don't exist, then do the normal checks which already existed before this change.
- If the `whitelistedApps` is empty / null / string, then continue with the checks which already existed before this change.
- Expectation is whitelistedApps should be an array we would check if the `appName` is there in any of the elements in the array. Expectation here is when the code will access `$userAttributes['guests']['whitelistedApps']` we should get a list of strings. 
- When `whitelistedApps` is an empty array then proceed ahead with further checks which were already present before this change.

sharidas commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,73 @@
+	 * @param array $attributes
+	 * @since 10.3.1
+	 */
+	public function __construct($event, $attributes) {
```
Removed these parameters.

jvillafanez commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,82 @@
+	private $attributes = [];
+
+	/**
+	 * Get the extended attributes of user for an app
```
To clarify, we're fetching user attributes, and we're asking for user attributes. I don't see why we need to know anything about apps here. As far as I know, the app is irrelevant.

sharidas replied:
Yes. I have removed the usage of apps or appnames from this code.

jvillafanez commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,82 @@
+	 * @param string $appName
+	 * @return mixed
+	 */
+	public function getAttribute($appName) {
```
As far as I know, this isn't needed. Keep things as easy as possible. Adding more methods will increase the confusion.
We only need one method for the listeners to add new user attributes, and another one for the `User` class to get those new attributes, nothing more.

jvillafanez commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,82 @@
+	 * @param string $attrubuteKey
+	 * @param mixed $attributeValue
+	 */
+	public function setAttributes($appName, $attrubuteKey, $attributeValue) {
```
As said, return true or false to give the apps a chance to know if the attribute has been set or not.
(btw, there is a typo in the "attrubuteKey")

sharidas replied:
The `setAttributes` method now returns true if it has set the attribute, else it returns false.
Fixed the typo.

jvillafanez commented:
One additional thing: the event is "owned" or "controlled" by the `User` class. This means that it's the `User` class the one that should decide the format of the event.

The `User` class doesn't need to know what app has filled the event, and it doesn't have any need for the app name. In addition, the `getExtendedAttributes` method is expected to return user attributes in a "attr => value" fashion. Returning app names isn't part of the specification.
The expectation is that the method will return something like:
```
"isGuest" => true,
"loginName" => "Joe_Joe",
"dateOfBirth" => "3/3/1954",
"blogSite" => "https://server/umpa/lumpa"
.....
```
Noone cares which app added the "blogSite" key.

sharidas commented on lib/private/App/AppManager.php:
```diff
@@ -191,9 +191,27 @@ public function isEnabledForUser($appId, $user = null) {
+		if ($user !== null) {
+			$userAppAttributes = $user->getExtendedAttributes();
+			/**
+			 * Check if the user is guest user, if so get the apps which it is allowed for.
```
Added it into the doc.

sharidas commented on lib/private/App/AppManager.php:
```diff
@@ -191,9 +191,27 @@ public function isEnabledForUser($appId, $user = null) {
+			 * the white listed apps to see if the appName is allowed for the user.
+			 * Else proceed with further checks in the method.
+			 */
+			if (\array_key_exists('guests', $userAppAttributes) &&
```
Removed the usage of  app name in the code. And updated to use `isset` instead of `array_key_exists`.

sharidas commented on lib/private/User/User.php:
```diff
@@ -113,6 +123,10 @@ public function __construct(Account $account, AccountMapper $mapper, $emitter =
 		if ($this->userSession === null) {
 			$this->userSession = \OC::$server->getUserSession();
 		}
+		if ($userExtendedAttributesEvent === null) {
```
Removed the event from the constructor. So new events are triggered each time, unless clear cache is set to false.

sharidas commented on lib/public/IUser.php:
```diff
@@ -238,4 +238,13 @@ public function setSearchTerms(array $terms);
+	 * @return array
+	 * @since 10.3.1
+	 */
+	public function getExtendedAttributes();
```
Included a `clearCache` parameter which is optional. By default the `clearCache` is assigned to `true`. Meaning if new event is triggered and updates the `userAttributes`. If we set the `clearCache` to false we would just return the existing content of `userAttributes`.

sharidas commented on lib/private/User/User.php:
```diff
@@ -555,4 +569,16 @@ public function setSearchTerms(array $terms) {
 	}
+
+	/**
+	 * get the attributes of user for apps
```
I have added documenation for this method. Let me know if this looks ok

jvillafanez commented on lib/private/User/User.php:
```diff
@@ -555,4 +562,38 @@ public function setSearchTerms(array $terms) {
+			return $this->userAttributes;
+		}
+
+		$userExtendedAttributesEvent = new UserExtendedAttributesEvent();
```
I think we need to send this user object in the event. I mean, the guests app should verify this user is a guest before adding the whitelistedApps; ldap would need something to join the ldap information with this user info...

We can send this object inside the event. This is a more flexible option taking into account that we'll need to send at least the uid and any app can fetch this same user via the user manager.

sharidas replied:
Updated the code. Now user object is passed through the event.

jvillafanez commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,70 @@
+	/**
+	 * Set the attributes
+	 * Here the attributes of the app with key and value are set.
+	 * The reason to use appName as one of the keys is its easier to know if the
```
no appName. Adjust the docs.

sharidas replied:
Verified and removed the appName in the file.

jvillafanez commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,70 @@
+	 * @return bool true if the attribute is set by the method false if the attribute is already present
+	 */
+	public function setAttributes($attributeKey, $attributeValue) {
+		if (isset($this->attributes[$attributeKey])) {
```
we need to handle the "null value" case.

sharidas replied:
Done

sharidas commented on lib/private/User/User.php:
```diff
@@ -555,4 +562,38 @@ public function setSearchTerms(array $terms) {
+	 */
+	public function getExtendedAttributes($clearCache = true) {
+		if (!$clearCache) {
+			return $this->userAttributes;
```
So with the updated code, the event is triggered if:
- `clearCache` is `true` or
- `userExtendedAttributes` is either `null` or empty array ( in this case even if the `clearCache` is `false`, the event will be triggered.

sharidas commented on lib/private/User/User.php:
```diff
@@ -555,4 +562,38 @@ public function setSearchTerms(array $terms) {
+	 * @return array
+	 * @since 10.3.1
+	 */
+	public function getExtendedAttributes($clearCache = true) {
```
Updated the `clearCache` to have default value `false`.

sharidas commented on lib/private/User/User.php:
```diff
@@ -77,6 +78,9 @@ class User implements IUser {
 	private $userSession;
 
+	/** @var array */
+	private $userAttributes = [];
```
Updated to `userExtendedAttributes`

sharidas commented on lib/private/App/AppManager.php:
```diff
@@ -191,9 +191,27 @@ public function isEnabledForUser($appId, $user = null) {
+			 * - If the list has some apps, only those apps will be available for that guest
+			 *
+			 */
+			if (isset($userAppAttributes['whiteListedApps'])) {
```
updated to `whitelistedAppsForGuests`.

jvillafanez commented:
I think we'll need to protect some methods to be called from the apps. It could be risky to let the apps modify things, and calling again the "getExtendedAttributes" could trigger an infinite loop.

The idea I have for this is to change all these "problematic" methods to check for a flag and throw an exception if the flag is set. This flag will be private, set as active before sending the event, and disable the flag after the sending the event.
I think this is an easy way to allow or deny what methods can be called inside the listeners.

sharidas commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,86 @@
+	 * @return bool true if the attribute is set by the method false if the attribute is already present
+	 */
+	public function setAttributes($attributeKey, $attributeValue) {
+		if (isset($this->attributes[$attributeKey]) && ($this->attributes[$attributeKey] !== null)) {
```
Updated code to use `isset` so that we can check if the index is there or not and if index is there and not null.

sharidas commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	/**
+	 * get the user who has triggered the event
```
Updated the doc

sharidas commented on lib/private/User/User.php:
```diff
@@ -555,4 +562,42 @@ public function setSearchTerms(array $terms) {
+		 * using empty because userExtendedAttributes could either be null or empty array
+		 * or an array of attributes
+		 */
+		if ($clearCache || empty($this->userExtendedAttributes)) {
```
Done. Instead of `empty`, check for null is added.

sharidas commented:
Well passing user object can be dangerous in one way, the apps can change user attributes through the user object. So lets say if the user id is passed, which is also dangerous because the app can access the same user object via userManager. I would like to grab inputs here, if we have to protect the set methods in the user object which can alter the user attributes? Another worrying factor here as @jvillafanez suggested in https://github.com/owncloud/core/pull/36258#issuecomment-541027084 if the app call `getExtendedAttributes` might cause an infinite loop?

I did an experiment using `clone` of user object, but unfortunately that does not help here. Because its a shallow copy and say if we try to change the display name, the change does reflect in the accounts table.

@micbar @PVince81 @VicDeo Let me know your thoughts regarding the same as this input will help us to move ahead with the pr.

jvillafanez commented on lib/private/User/User.php:
```diff
@@ -140,8 +158,13 @@ public function getUserName() {
 	 */
 	public function setUserName($userName) {
+		if (!$this->allowUserAccountUpdate) {
+			throw new NotPermitUserAccountException("Operation cannot be allowed as other apps are fetching extended attributes of this user.");
```
Note that this needs a change in the interface

sharidas replied:
Updated the interface.

jvillafanez commented on lib/public/NotPermitUserAccountException.php:
```diff
@@ -0,0 +1,42 @@
+ * @package OCP
+ * @since 10.3.1
+ */
+class NotPermitUserAccountException extends \Exception {
```
Needs a better naming.
If this is intended to be used only for this specific case, `NotPermittedUserAction` seems a better choice. You can also put it inside the public/User folder.
If this could be used in a more generic way, `NotPermittedAction` could also fit.

sharidas replied:
Done

jvillafanez commented on lib/private/User/User.php:
```diff
@@ -555,4 +609,55 @@ public function setSearchTerms(array $terms) {
+			 * user.
+			 */
+			$this->allowUserAccountUpdate = false;
+			$this->eventDispatcher->dispatch(UserExtendedAttributesEvent::USER_EXTENDED_ATTRIBUTES, $userExtendedAttributesEvent);
```
we need to handle exceptions triggered from inside the listeners, specially because the flag could remain active.

sharidas replied:
I have updated the code to handle the exception thrown. Also I have updated the AppManager where `getExtendedAttributes` is called. I have added try/catch block. The catch block is kept empty. So that the code flow will happen as if its a regular user and not a guest user.

jvillafanez commented on tests/lib/User/UserExtendedAttributesEventTest.php:
```diff
@@ -0,0 +1,78 @@
+	 * This test is to get the user whose extended attributes are to be fetched
+	 */
+	public function testGetUser() {
+		$this->assertEquals($this->userExtendedAttributesEvent->getUser(), $this->user);
```
`assertSame` might be a better choice because we want to ensure it's the same instance, not a copy or a difference object with the same attributes.

jvillafanez commented on tests/lib/User/UserExtendedAttributesEventTest.php:
```diff
@@ -0,0 +1,78 @@
+	 * @param bool $expectedResult
+	 */
+	public function testSetAttributes($attributeKey, $attributeVal, $expectedResult) {
+		if ($attributeKey === 'one' && $expectedResult === false) {
```
I don't understand this "if"

sharidas replied:
I have added the comment for this before this condition. So what I am trying to test here is:
- add attribute with key `one`.
- Again try to set the attribute with key `one`
- `setAttributes` should return false.

jvillafanez replied:
this should be removed now that we have the other test

sharidas replied:
Removed this test.

jvillafanez commented on tests/lib/User/UserExtendedAttributesEventTest.php:
```diff
@@ -0,0 +1,78 @@
+		$this->assertEquals($this->userExtendedAttributesEvent->setAttributes($attributeKey, $attributeVal), $expectedResult);
+	}
+
+	public function testGetAttributes() {
```
I think "testSetGetAttributes" is more aligned with what the test is doing.

sharidas replied:
Updated.

jvillafanez commented on tests/lib/User/UserTest.php:
```diff
@@ -544,4 +545,25 @@ public function setTermsData() {
+			->method('dispatch')
+			->with(UserExtendedAttributesEvent::USER_EXTENDED_ATTRIBUTES, $userExtendedAttributesEvent);
+
+		$this->assertNull($this->user->getExtendedAttributes($clearCache));
```
```
 * @param bool $clearCache, set to true if user attributes should be created every time, else false is set to reuse the userExtendedAttributes cache.
	 * @return array
	 * @since 10.3.1
	 */
	public function getExtendedAttributes($clearCache = false);
```
The function always returns an array. This test is invalid.

sharidas replied:
Done

jvillafanez commented:
BTW, we'll need tests for the changes made in the `User` class

sharidas commented:
> BTW, we'll need tests for the changes made in the `User` class

I have updated the tests for the exceptions thrown for the `User` class.

jvillafanez commented on tests/lib/User/UserTest.php:
```diff
@@ -544,4 +546,103 @@ public function setTermsData() {
+			->with(UserExtendedAttributesEvent::USER_EXTENDED_ATTRIBUTES, $userExtendedAttributesEvent)
+			->willThrowException(new NotPermittedActionException());
+
+		$this->assertEquals($this->user->getExtendedAttributes(), []);
```
This makes things confusing: either you assert the call throws an exception, or you assert that the method returns an empty array.

sharidas replied:
That was a mistake. Fixed it.

jvillafanez commented on lib/public/User/UserExtendedAttributesEvent.php:
```diff
@@ -0,0 +1,98 @@
+class UserExtendedAttributesEvent extends Event {
+	const USER_EXTENDED_ATTRIBUTES = 'UserExtendedAttributesEvent';
+
+	/** @var array|null */
```
adjust doc here

sharidas replied:
Done

sharidas commented on lib/private/User/User.php:
```diff
@@ -555,4 +618,62 @@ public function setSearchTerms(array $terms) {
+			try {
+				$this->eventDispatcher->dispatch(UserExtendedAttributesEvent::USER_EXTENDED_ATTRIBUTES, $userExtendedAttributesEvent);
+			} catch (NotPermittedActionException $ex) {
+				$this->logger->logException($ex);
```
Removed the DI for logger. Also removed the catch block. Reduced to try/finally block.

sharidas commented:
- The `getExtendedAttributes` is now protected with the flag. So that the listeners cannot call this method, which might cause infinite loop.
- Updated the unit test for the flag to test
   - normal execution. That is after the normal execution the flag is reset to `true`.
   - With exception `NotPermittedActionException`. After this exception is thrown, verified that the value of flag is set to `true`.
   - Tested with generic exception `Exception`. After this exception is thrown, verified that the value of the flag is set to `true`.

- With the current approach the event does not reach the other listeners if the initial/first listener throws the exception. So for instance I added a test code in the comments app ( as this app is found first in the list ), and tried to set the display name. I observed that code did not reached `guests` listener. 
    - One solution I see is to catch the exception `NotPermittedActionException` in the `getExtendedAttributes()` and then just log it inside the catch block. I think for this method we can catch the `NotPermittedActionException` exception. But if any other exception comes then lets allow the exception to bubble up? Or another way is we can ask the listeners to catch the exception? Even that way it should work imho. This second solution is like completely relying on apps. Personally I am in favour of logging the exception of `NotPermittedActionException` in the `User` class and then move ahead. And for any other exceptions let the listener handle it.

jvillafanez commented:
> With the current approach the event does not reach the other listeners if the initial/first listener throws the exception. So for instance I added a test code in the comments app ( as this app is found first in the list ), and tried to set the display name. I observed that code did not reached guests listener.

Looks like it isn't possible to force the event to reach all the listeners. We'd need a custom eventDispatcher (which is out of scope at the moment).
So, for the `getExtendedAttributes` method, I think the best solution is to let the exceptions (even the `NotPermittedActionException`) bubble up out of the method because it can't handle those. We'll have to rely on the listeners not to leak any exception.

jvillafanez commented on lib/private/User/User.php:
```diff
@@ -555,4 +611,65 @@ public function setSearchTerms(array $terms) {
+		 * using empty because userExtendedAttributes could either be null or empty array
+		 * or an array of attributes
+		 */
+		if ($clearCache || !isset($this->userExtendedAttributes) || [] === $this->userExtendedAttributes) {
```
We don't need to retrigger the event if the `userExtendedAttributes` is an empty array.
You triggered the event, no one listened and you get an empty array. This is a valid result. There is no need to trigger the event again to get another empty array.

jvillafanez commented:
Hopefully last change. Please make sure the comments are updated and match the code behaviour.

PVince81 commented:
I don't feel comfortable with this approach, it seems it doesn't fully suit.

Looking at the guests app we are setting an attribute "whitelistedAppsForGuests" which sounds like a generic setting, not something specific to a user. So having this in a user specific attribute feels wrong.

Thinking further: if the user attribute is something that defines what is accessible for a user, then maybe it fits more into a "roles" concept or "permissions" concept, and in this case the attribute would need to be called "whitelistedApps", as it applies only for this one user.

Maybe we should rather introduce the concept of "app access" into core in a more API-like way instead of just attributes. We already have the concept of "enabled apps for users" but here it's a softer one as the app needs to be loaded and its elements need to be discarded.

In a broader view, an app should have a way to find out whether the current user is supposed to see its controls or not, so having an API for "app access" or "app permissions" might be more suitable.

@DeepDiver1975 @micbar any thoughts on this ? does this fit somehow into the discussions related to roles concept ?

jvillafanez commented:
The problem we've noticed at the last moment is that the all the apps are required to be loaded before calling the `getExtendedAttributes` method.
For this particular case, the guests app is required to be loaded so we can know what other apps are whitelisted. In the worst case the guests app is loaded the last, we might wrongly assume that the rest of the apps can be loaded and be accessed normally by that user.

I think the idea is worth to keep because it has other uses such as properly expose the login name of the user, expose LDAP attributes for LDAP users, or other information for shibboleth or openIdConnect.

> Looking at the guests app we are setting an attribute "whitelistedAppsForGuests" which sounds like a generic setting, not something specific to a user. So having this in a user specific attribute feels wrong

The original idea was to use 2 keys: the "isGuest" in order to identify guests users, and the  "whitelistedApps" to include the list. "whitelistedAppsForGuests" was a simple way to merge both keys since the intention was to use that key for both identify a guest user and get the list of apps.

In any case, I'm starting to think this approach won't work for guests. Guests are added in the DB backend and they aren't inside a "guests" backend part of the guests app. When the guests app is disabled, guests users will still be available, but we won't be able to identify them as a guest with this approach because the app will be disabled.

PVince81 commented:
didn't we have an event (hook) for the moment after which all apps have finished loading ?

sharidas commented:
> didn't we have an event (hook) for the moment after which all apps have finished loading ?

Searched the code base in core, and found this https://github.com/owncloud/core/blob/master/lib/private/legacy/app.php#L162
And this would return false https://github.com/owncloud/core/blob/master/lib/private/legacy/hook.php#L91

Other than this I couldn't find further references.

mmattel commented:
@phil-davis @sharidas is there any progress?

mmattel commented:
@jvillafanez maybe you could take a look on to bring this forward...

jvillafanez commented:
Looks good to me. Maybe @DeepDiver1975 can have a second look for the approach.
We might need to rebase the branch and retest.

phil-davis commented:
> We might need to rebase the branch and retest.

Rebased in #40257 - we can review and comment there, and if needed and good it can easily be merged from there.

phil-davis commented:
PR #40257 is passing now. I had to just fixup some code-style and unit test things. So someone can comment there if this is wanted, and if the code is OK.

phil-davis commented:
This code has been merged in PR #40257
